### PR TITLE
Add new method to RealityDataSource

### PR DESCRIPTION
**Pull request by MarcBedard8**
Gets spatial location and extents of this reality data source.
Gets information to identify the product and engine that create this reality data.
Split realityDataSource implementation based on providers.

### CODE REVIEW

File: core/frontend/src/RealityDataSource.ts [hunk 1]
```diff
@@ -5,15 +5,44 @@
...
+/** This interface provide spatial location and volume of interest, in meters, centered around `spatial location`
+ * @alpha
+ */
+export interface SpatialLocationAndExtents {
+  /** location of the point at the center of the reaity data */
+  location: Cartographic | EcefLocationProps;
+  /** extents of the volume around location */
+  worldRange: Range3dProps;
+  /** true if this reality data is geolocated */
+  isGeolocated: boolean;
```
Comment (pmconne) [inline core/frontend/src/RealityDataSource.ts:28]: Under what circumstances would `isGeolocated` be false? Doesn't the existence of a location and worldRange imply it's geolocated?

Comment (MarcBedard8) [reply core/frontend/src/RealityDataSource.ts:28]: Some dataset that came from ContextCapture are in a custom coordinate system for which we cannot compute valid lat / long / elevation. Nothing in the dataset indicate that, thus when we found a Reality Data that is more than 5 Km below the ground, we mark it as not geolocated. For these dataset, we can still compute location and worldRange and we use these info to create a blank iModelConnection, so we can display them.

Comment (pmconne) [reply core/frontend/src/RealityDataSource.ts:28]: See my edit above. Can you clarify documentation please?

Comment (MarcBedard8) [reply core/frontend/src/RealityDataSource.ts:32]: Push updated doc.

File: core/frontend/src/RealityDataSource.ts [hunk 2]
```diff
@@ -5,15 +5,48 @@
...
+ */
+export interface PublisherProductInfo {
+  /** product that create this reality data */
+  product: string;
+  /** engine that create this reality data */
+  engine: string;
+  /** the version of the engine that create this reality data */
+  version: string;
+  /** the number of children at the root of this reality data */
+  rootChildren?: string;
```
Comment (pmconne) [inline core/frontend/src/RealityDataSource.ts:46]: This is "the number of children" but its type is `string`. What does this have to do with "publisher product info"?

File: core/frontend/src/tile/OPCFormatInterpreter.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import { Cartographic, EcefLocation, EcefLocationProps } from "@itwin/core-common";
+import { Range3d, Range3dProps } from "@itwin/core-geometry";
+import { ALong, CRSManager, Downloader, OnlineEngine, OPCReader, OrbitGtBounds, PageCachedFile, PointCloudReader, UrlFS } from "@itwin/core-orbitgt";
+import { FrontendLoggerCategory } from "../FrontendLoggerCategory";
+import { DownloaderNode } from "@itwin/core-orbitgt/lib/cjs/system/runtime/DownloaderNode";
+import { Logger } from "@itwin/core-bentley";
+
+const loggerCategory: string = FrontendLoggerCategory.RealityData;
+
+interface SpatialLocationAndExtents {
```
Comment (pmconne) [inline core/frontend/src/tile/OPCFormatInterpreter.ts:15]: This is a duplication of the same interface exported elsewhere.

File: core/frontend/src/RealityDataSource.ts [hunk 4]
```diff
@@ -5,15 +5,48 @@
...
-/**
- * This interface provide methods used to access a reality data from a reality data provider
+const loggerCategory: string = FrontendLoggerCategory.RealityData;
+
+/** This interface provide spatial location and volume of interest, in meters, centered around `spatial location`
+ * @alpha
+ */
+export interface SpatialLocationAndExtents {
+  /** location of the point at the center of the reaity data */
+  location: Cartographic | EcefLocationProps;
```
Comment (pmconne) [inline core/frontend/src/RealityDataSource.ts:24]: Is this interface a "props" (JSON) type or not? It mixes EcefLocationProps and Range3dProps, which are JSON types; with Cartographic, which is a class type.
It seems to me it should use EcefLocation and Range3d instead.

File: core/frontend/src/tile/ThreeDTileFormatInterpreter.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  /** Gets information to identify the product and engine that create this reality data
+   * Will return undefined if cannot be resolved
+   * @param rootDocjson root document file in json format
+   * @returns information to identify the product and engine that create this reality data
+   * @alpha
+   */
+  public static getPublisherProductInfo(rootDocjson: any): PublisherProductInfo {
+    const info: PublisherProductInfo = {product: "", engine: "", version: "", rootChildren: ""};
+    if (rootDocjson && rootDocjson.root) {
+      info.rootChildren = rootDocjson.root.children.length;
```
Comment (pmconne) [inline core/frontend/src/tile/ThreeDTileFormatInterpreter.ts:96]: I assume `root.children` is an array and therefore you are assigning a number to a `string` property.
Compiler can't tell you that because `any`.

File: core/frontend/src/tile/ThreeDTileFormatInterpreter.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        info.product = rootDocjson.root.SMPublisherInfo.Product ? rootDocjson.root.SMPublisherInfo.Product : "";
+        info.engine =  rootDocjson.root.SMPublisherInfo.Publisher ? rootDocjson.root.SMPublisherInfo.Publisher : "";
+        info.version = rootDocjson.root.SMPublisherInfo["Publisher Version"] ? rootDocjson.root.SMPublisherInfo["Publisher Version"] : "" ;
+      }
+    }
+    return info;
+  }
+}
+
+class RealityModelTileUtils {
```
Comment (pmconne) [inline core/frontend/src/tile/ThreeDTileFormatInterpreter.ts:107]: nitpick but this namespace serves no purpose.

File: core/frontend/src/RealityDataSource.ts (hunk 2 above)
Comment (MarcBedard8) [reply core/frontend/src/RealityDataSource.ts]: You are right. I've moved rootChildren from PublisherProductInfo to internal ThreeDTileFileInfo.
Change rootChildren type to number.
Add a new internal method on ThreeDTileFormatInterpreter to getFileInfo.

File: core/frontend/src/tile/OPCFormatInterpreter.ts (hunk 3 above)
Comment (MarcBedard8) [reply core/frontend/src/tile/OPCFormatInterpreter.ts]: Remove duplicate SpatialLocationAndExtents interface definition.

File: core/frontend/src/tile/ThreeDTileFormatInterpreter.ts (hunk 5 above)
Comment (MarcBedard8) [reply core/frontend/src/tile/ThreeDTileFormatInterpreter.ts]: Change rootChildren type to number and use parseInt.

File: core/frontend/src/tile/ThreeDTileFormatInterpreter.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  /** Gets information about 3dTile file for this reality data
+   * Will return undefined if cannot be resolved
+   * @param rootDocjson root document file in json format
+   * @returns information about 3dTile file for this reality data
+   * @internal
+   */
+  public static getFileInfo(rootDocjson: any): ThreeDTileFileInfo {
+    const info: ThreeDTileFileInfo = { rootChildren: 0};
+    if (rootDocjson && rootDocjson.root) {
+      info.rootChildren = parseInt(rootDocjson.root.children.length, 10);
```
Comment (pmconne) [inline core/frontend/src/tile/ThreeDTileFormatInterpreter.ts:120]: @MarcBedard8 Is `rootDocjson.root.children` an array? If so, `parseInt` is going to produce a run-time exception, because it expects a string.

File: core/frontend/src/RealityDataSource.ts (hunk 4 above)
Comment (MarcBedard8) [reply core/frontend/src/RealityDataSource.ts]: SpatialLocationAndExtents is not a Props type. Change member to use class type instead of props.

File: core/frontend/src/tile/ThreeDTileFormatInterpreter.ts (hunk 6 above)
Comment (MarcBedard8) [reply core/frontend/src/tile/ThreeDTileFormatInterpreter.ts]: Remove namespace RealityModelTileUtils and add method to ThreeDTileFormatInterpreter class instead.

File: core/frontend/src/tile/ThreeDTileFormatInterpreter.ts (hunk 7 above)
Comment (pmconne) [reply core/frontend/src/tile/ThreeDTileFormatInterpreter.ts]: You're still not checking if `children` is `undefined`.
Change it to: 
```
const info: ThreeDTileFileInfo = {
  rootChildren: rootDocjson?.root?.children?.length ?? 0,
};
```

Comment (MarcBedard8) [reply core/frontend/src/tile/ThreeDTileFormatInterpreter.ts]: Done

**Review by pmconne — APPROVED** [review]
